lrauv_ignition_plugins/src/ScienceSensorsSystem.cc

```diff
@@ -612,39 +701,304 @@ void ScienceSensorsSystemPrivate::GenerateOctrees()
+      << " for trilinear interpolation." << std::endl;
+  }
+
+  // Step 2: exclude z slice of 1st NN from further searches.
+
+  // Remove all points in the z slice of the 1st NN, so that the 2nd NN will be
+  // found in another z slice.
+  // Take the set difference between original point cloud and the point cloud
+  // of the z slice of the 1st NN, to get the remaining points not on the 1st z
+  // slice.
+  std::vector<pcl::PointXYZ> ptsExceptZSlice1;
+  std::set_difference(
+    this->timeSpaceCoords[this->timeIdx]->points.begin(),
+    this->timeSpaceCoords[this->timeIdx]->points.end(),
+    zSlice1.points.begin(),
+    zSlice1.points.end(),
+    std::inserter(ptsExceptZSlice1, ptsExceptZSlice1.begin()),
+    std::bind(&ScienceSensorsSystemPrivate::comparePclPoints, this,
+      std::placeholders::_1, std::placeholders::_2));
+
+  // Step 3: Look for 2nd NN everywhere except z slice of 1st NN.
+  // In this 2nd z-slice, search for 4 NNs
+
+  // Create point cloud from set difference result
+  pcl::PointCloud<pcl::PointXYZ> cloudExceptZSlice1;
+  cloudExceptZSlice1.points.insert(cloudExceptZSlice1.points.begin(),
+    ptsExceptZSlice1.begin(), ptsExceptZSlice1.end());
+
+  // Create octree
+  auto octreeExceptZSlice1 = pcl::octree::OctreePointCloudSearch<pcl::PointXYZ>(
+    this->spatialRes);
+  octreeExceptZSlice1.setInputCloud(cloudExceptZSlice1.makeShared());
+  octreeExceptZSlice1.addPointsFromInputCloud();
+
+  // Search results
+  std::vector<int> inds2;
+  std::vector<float> sqrDists2;
+  int nnIdx2 = -1;
+  float minDist2 = -1.0f;
+  float nnZ2 = -1.0f;
+
+  // Look for 2nd NN
+  if (octreeExceptZSlice1.getLeafCount() > 0)
+  {
+    if (octreeExceptZSlice1.nearestKSearch(_pt, 1, inds2, sqrDists2)
+      <= 0)
+    {
+      ignwarn << "Cannot find 2nd NN for trilinear interpolation" << std::endl;
+      return;
+    }
+    else
+    {
+      // Take closest point as 2nd NN
+      nnIdx2 = inds2[0];
+      minDist2 = sqrDists2[0];
+      // Get z of neighbor
+      nnZ2 = cloudExceptZSlice1.at(nnIdx2).z;
+
+      igndbg << "Slice 2 nnIdx " << nnIdx2 << ", dist " << sqrt(minDist2)
+        << std::endl;
+    }
+  }
+
+  // Step 4: Look for 4 NNs in the z slice of 2nd NN
+
+  // Get raw Eigen matrix
+  // Matrix size 3 x n
+  Eigen::MatrixXf matrixExceptZSlice1 =
+    cloudExceptZSlice1.getMatrixXfMap(3, 4, 0).block(
+      0, 0, 3, cloudExceptZSlice1.size());
+
+  // Search in z slice of 1st NN for 4 nearest neighbors in this slice
+  this->SearchInDepthSlice(_pt, nnZ2, matrixExceptZSlice1, zSlice2, zSliceInds2,
+    interpolatorInds2, interpolatorSqrDists2, _interpolators2);
+  if (interpolatorInds2.size() < 4 || interpolatorSqrDists2.size() < 4)
+  {
+    ignwarn << "Could not find enough neighbors in 2nd slice z = " << nnZ
+      << " for trilinear interpolation." << std::endl;
+  }
+}
+
+/////////////////////////////////////////////////
+void ScienceSensorsSystemPrivate::SearchInDepthSlice(
```

Comment:
I would suggest separating the slice generation from the "SearchInSlice" method. 

The slices can be generated separately prior to running the search and then passed into the search function instead of returning the sliced cloud from the first search and generating a new slice for the second search.

Reply:
Part of 7c85554